### ISSUE TITLE
RangeError: Maximum call stack size exceeded at recursiveFindProvider

### ISSUE THREAD

**Issue by pheinicke** [getsaf/shallow-render#176]
Hi!

When updating from shallow-render 10.0.1 to 10.1.0 I'm having trouble with some failing tests that worked fine before. Only tests for components that open a MatDialog (from https://material.angular.io/) fail.

```
 FAIL  src/app/shared/_components/confirm-dialog/confirm-dialog.component.spec.ts (15.045 s)
  ● ConfirmDialogComponent › should create

    RangeError: Maximum call stack size exceeded
        at Array.find (<anonymous>)

      at recursiveFindProvider (node_modules/shallow-render/lib/tools/mock-provider.ts:7:12)
```

The component itself is pretty simple:
```
export interface ConfirmDialogData {
    title: string;
    message: string;
    confirm?: string;
    dismiss?: string;
}

@Component({
    selector: 'app-confirm-dialog',
    templateUrl: './confirm-dialog.component.html',
    styleUrls: ['./confirm-dialog.component.scss'],
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class ConfirmDialogComponent {
    title: string;
    message: string;
    confirm: string;
    dismiss: string;

    constructor(@Inject(MAT_DIALOG_DATA) data: ConfirmDialogData) {
        this.title = data.title;
        this.message = data.message;
        this.confirm = data.confirm || 'common.confirm';
        this.dismiss = data.dismiss || 'common.dismiss';
    }

    static open(dialog: MatDialog, data: ConfirmDialogData): MatDialogRef<ConfirmDialogComponent> {
        return dialog.open(ConfirmDialogComponent, { data });
    }
}

export function onConfirm(dialog: MatDialog, data: ConfirmDialogData): Observable<void> {
    return ConfirmDialogComponent.open(dialog, data)
        .afterClosed()
        .pipe(
            filter((confirmed) => confirmed),
            mapTo(undefined)
        );
}
```

The test:
```
describe('ConfirmDialogComponent', () => {
    let shallow: Shallow<ConfirmDialogComponent>;

    beforeEach(() => {
        shallow = new Shallow(ConfirmDialogComponent, SharedModule).provide([
            { provide: MAT_DIALOG_DATA, useValue: { title: 'title', message: 'message' } }
        ]);
    });

    it('should create', async () => {
        const { fixture } = await shallow.render();
        expect(fixture).toBeTruthy();
    });
});
```

Any help would be appreciated!

**Comment by getsaf** [comment]
Thanks, there was a change in the provider logic where mocked components now have their component-level providers mocked. Will search try to reproduce and get a fix in ASAP.

**Comment by getsaf** [comment]
Just a note, the code you provided was perfect for helping reproduce/debug the issue, thanks for that.

This was indeed a regression in mocking providers that caused infinite recursion for certain module structures.

I just released [v10.1.1](https://github.com/getsaf/shallow-render/releases/tag/v10.1.1) with the fix.

**Comment by pheinicke** [comment]
Works like a charm! Thanks for your quick response!
